File: src/github.com/oniony/TMSU/cli/info.go
// Copyright 2011-2017 Paul Ruane.

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package cli

import (
	"fmt"
	"github.com/oniony/TMSU/common/terminal/ansi"
	"github.com/oniony/TMSU/storage"
	"math"
	"os"
	"strconv"
)

var InfoCommand = Command{
	Name:        "info",
	Synopsis:    "Show database information",
	Usages:      []string{"tmsu info"},
	Description: "Shows the database information.",
	Options: Options{
		Option{"--stats", "-s", "show statistics", false, ""},
		Option{"--usage", "-u", "show tag usage breakdown", false, ""}},
	Exec:    infoExec,
	Aliases: []string{"stats"},
}

// unexported

func infoExec(options Options, args []string, databasePath string) (error, warnings) {
	stats := options.HasOption("--stats")
	usage := options.HasOption("--usage")
	colour, err := useColour(options)
	if err != nil {
		return err, nil
	}

	store, err := openDatabase(databasePath)
	if err != nil {
		return err, nil
	}
	defer store.Close()

	tx, err := store.Begin()
	if err != nil {
		return err, nil
	}
	defer tx.Commit()

	showBasic(store, tx, colour)

	if stats {
		showStatistics(store, tx, colour)
	}
	if usage {
		showUsage(store, tx, colour)
	}

	return nil, nil
}

func showBasic(store *storage.Storage, tx *storage.Tx, colour bool) error {
	printInfo("Database", store.DbPath, colour)
	printInfo("Root path", store.RootPath, colour)

	stat, err := os.Stat(store.DbPath)
	if err != nil {
		return err
	}

	printInfo("Size", stat.Size(), colour)

	return nil
}

func showStatistics(store *storage.Storage, tx *storage.Tx, colour bool) error {
	tagCount, err := store.TagCount(tx)
	if err != nil {
		return fmt.Errorf("could not retrieve tag count: %v", err)
	}

	valueCount, err := store.ValueCount(tx)
	if err != nil {
		return fmt.Errorf("could not retrieve value count: %v", err)
	}

	fileCount, err := store.FileCount(tx)
	if err != nil {
		return fmt.Errorf("could not retrieve file count: %v", err)
	}

	fileTagCount, err := store.FileTagCount(tx)
	if err != nil {
		return fmt.Errorf("could not retrieve taggings count: %v", err)
	}

	var averageTagsPerFile float32
	if fileCount > 0 {
		averageTagsPerFile = float32(fileTagCount) / float32(fileCount)
	}

	var averageFilesPerTag float32
	if tagCount > 0 {
		averageFilesPerTag = float32(fileTagCount) / float32(tagCount)
	}

	fmt.Println()
	printInfo("Tags", tagCount, colour)
	printInfo("Values", valueCount, colour)
	printInfo("Files", fileCount, colour)
	printInfo("Taggings", fileTagCount, colour)
	printInfof("Mean tags per file", "%1.2f", averageTagsPerFile, colour)
	printInfof("Mean files per tag", "%1.2f", averageFilesPerTag, colour)

	return nil
}

func showUsage(store *storage.Storage, tx *storage.Tx, colour bool) error {
	tagUsages, err := store.TagUsage(tx)
	if err != nil {
		return fmt.Errorf("could not retrieve tag usage: %v", err)
	}

	maxLength := 0
	maxCountWidth := 0

	for _, tagUsage := range tagUsages {
		countWidth := int(math.Log(float64(tagUsage.FileCount)))
		if countWidth > maxCountWidth {
			maxCountWidth = countWidth
		}
		if len(tagUsage.Name) > maxLength {
			maxLength = len(tagUsage.Name)
		}
	}

	fmt.Println()
	for _, tagUsage := range tagUsages {
		fileCount := strconv.FormatUint(uint64(tagUsage.FileCount), 10)
		if colour {
			fileCount = ansi.Yellow(fileCount)
		}

		fmt.Printf("  %*s %*v\n", -maxLength, tagUsage.Name, maxCountWidth, fileCount)
	}

	return nil
}

func printInfo(name string, value interface{}, colour bool) {
	printInfof(name, "%v", value, colour)
}

func printInfof(name, format string, value interface{}, colour bool) {
	if colour {
		format = ansi.Green(format)
	}

	fmt.Printf("%v: "+format+"\n", name, value)
}

File: src/cli/info.rs
use structopt::StructOpt;

use std::fmt;

use termion::color;

use crate::api;
use crate::errors::*;

/// Shows database information
#[derive(Debug, StructOpt)]
pub struct InfoOptions {
    /// Shows statistics
    #[structopt(short, long)]
    stats: bool,

    /// Shows tag usage breakdown
    #[structopt(short, long)]
    usage: bool,
}

impl InfoOptions {
    pub fn execute(&self, global_opts: &super::GlobalOptions) -> Result<()> {
        let db_path = super::locate_db(&global_opts.database)?;
        info!("Database path: {}", db_path.display());

        let use_colors = super::should_use_colour(&global_opts.color);

        let info_output = api::info::run_info(&db_path, self.stats, self.usage)?;
        print_output(&info_output, use_colors);

        Ok(())
    }
}

fn print_output(info_data: &api::info::InfoOutput, use_colors: bool) {
    print_entry("Database", &info_data.db_path.display(), use_colors);
    print_entry("Root path", &info_data.root_path.display(), use_colors);
    print_entry("Size", &info_data.size, use_colors);

    if let Some(stats_data) = &info_data.stats_info {
        print_stats(stats_data, use_colors);
    }

    if let Some(usage_data) = &info_data.usage_info {
        print_usage(usage_data, use_colors);
    }
}

fn print_stats(stats_data: &api::info::StatsOutput, use_colors: bool) {
    println!();

    print_entry("Tags", &stats_data.tag_count, use_colors);
    print_entry("Values", &stats_data.value_count, use_colors);
    print_entry("Files", &stats_data.file_count, use_colors);
    let taggings = stats_data.file_tag_count;
    print_entry("Taggings", &taggings, use_colors);

    let average_tags_per_file = taggings as f64 / stats_data.file_count as f64;
    print_entry(
        "Mean tags per file",
        &format!("{:.2}", average_tags_per_file),
        use_colors,
    );
    let average_files_per_tag = taggings as f64 / stats_data.tag_count as f64;
    print_entry(
        "Mean files per tag",
        &format!("{:.2}", average_files_per_tag),
        use_colors,
    );
}

fn print_usage(usage_data: &api::info::UsageOutput, use_colors: bool) {
    let max_tag_width = usage_data
        .tag_data
        .iter()
        .map(|tag_info| tag_info.name.len())
        .max()
        .unwrap_or_default();

    println!();
    for tag_info in &usage_data.tag_data {
        print_tag_info(tag_info, max_tag_width, use_colors);
    }
}

fn print_entry<T: fmt::Display>(key: &str, value: &T, use_colors: bool) {
    if use_colors {
        println!(
            "{}: {}{}{}",
            key,
            color::Fg(color::Green),
            value,
            color::Fg(color::Reset)
        );
    } else {
        println!("{}: {}", key, value);
    }
}

fn print_tag_info(tag_info: &api::info::TagInfo, max_tag_width: usize, use_colors: bool) {
    if use_colors {
        println!(
            "  {:width$} {}{}{}",
            &tag_info.name,
            color::Fg(color::Yellow),
            tag_info.count,
            color::Fg(color::Reset),
            width = max_tag_width
        );
    } else {
        println!(
            "  {:width$} {}",
            &tag_info.name,
            tag_info.count,
            width = max_tag_width
        );
    }
}
